src/celeritas/user/detail/StepDiagnosticImpl.hh

```diff
@@ -0,0 +1,57 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/user/detail/StepDiagnosticImpl.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/math/Algorithms.hh"
+#include "corecel/math/Atomics.hh"
+#include "celeritas/global/CoreTrackView.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Collect distribution of steps per track for each particle type.
+ */
+inline CELER_FUNCTION void
+tally_steps(CoreTrackView const& track,
+            NativeCRef<ParticleTallyParamsData>&& params,
+            NativeRef<ParticleTallyStateData>&& state)
```

Comment:
```suggestion
            NativeCRef<ParticleTallyParamsData>const& params,
            NativeRef<ParticleTallyStateData>const& state)
```
These should be `const&`: the semantics are a little different between `Obj&&` and `template<T> ... T&&`: because the latter can match `const&` and allow you to pass const data, whereas here you are requiring the data to be "moved" into the function. Once you make this change you will be able to declare the kernel arguments as `const`.

Reply:
Gotcha, thanks!